Usprawnienia UART

marcinn opened the pull request:
Dzień dobry, panie Milek7. Podsyłam propozycję usprawnień obsługi UART.

1. UI / Debug Panel: dodana sekcja UART z możliwością zmiany portu, baud rate, wstrzymania transmisji i z krótkim podsumowaniem stanu transmisji
2. Korekty obsługi UART, dzięki którym można podłączyć i odłączyć interfejs w trakcie działania symulatora.

Dzięki tym zmianom o wiele łatwiej prowadzić jest prace nad sprzętowymi pulpitami (można przełączać się między płytką prototypową, można zatrzymać transmisję i wgrać nowszy firmware bez wyłączania symulacji), oraz w razie przypadkowego rozłączenia komunikacji można bez problemu kontynuować jazdę bez pulpitu lub po jego ponownym połączeniu.

Uwagi:
- kompilację robiłem na linuxie, niestety nie miałem możliwości sprawdzić na Windowsie
- lista portów pobierana jest przez sp_list_ports() z libserialport. Na linuxie działa, na Windowsie nie miałem jak sprawdzić. EDIT: widzę automat CI - super!

![image](https://user-images.githubusercontent.com/139032/138983262-919fa773-28a8-47c9-a2ca-f05af3eac440.png)


PS. Odrębny PR przygotowany jest dla pana TMJ: https://github.com/tmj-fstate/maszyna/pull/20

marcinn committed:
(uart) fix uart synchronization routine

marcinn committed:
(uart, improvement) not quitting when serial port is unavailable + automatic reconnect + no repeating uart error messages

marcinn committed:
(uart, debug) add UART debug panel with connection information

marcinn committed:
(uart) add possibility to enable/disable UART transmission from Debug panel

marcinn committed:
(uart) add serial ports enumeration and possibility to change at runtime

marcinn committed:
(uart) fix port change handling, add packet stats resetting

marcinn committed:
(uart) change incorrect name "enumeration" -> "port finding"

marcinn committed:
(uart, driveruipanels) change uart port selection widget from listbox to combo

marcinn committed:
(uart, driveruipanels) add possibility to change uart baud rate in debug panel

marcinn commented:
Checks wywaliły się na Ubuntu, bo obraz Ubuntu 16.04 x86-64 nie jest już dostępny.

Milek7 commented on application.h:
```diff
@@ -91,6 +94,9 @@ class eu07_application {
     bool
         is_client() const;
+#ifdef WITH_UART
+    UartStatus uart_status;
```
Tak trochę średnio to pasuje. Myślę że lepiej żeby to siedziało w klasie uart_input.

marcinn replied:
Ok.

Milek7 commented on driveruilayer.cpp:
```diff
@@ -40,6 +40,9 @@ driver_ui::driver_ui() {
 
     m_aidpanel.title = STR("Driving Aid");
 
+    m_scenariopanel.size_min = { 435, 50 };
```
Pomyłka z copypastą?

marcinn replied:
Copy+paste+skleroza. Poprawię.

Milek7 commented on driveruipanels.cpp:
```diff
@@ -577,6 +602,13 @@ debug_panel::render() {
     ImGui::PopFont();
 }
 
+#ifdef WITH_UART
```
Wygląda na niepotrzebne.

Milek7 commented on driveruipanels.h:
```diff
@@ -107,7 +113,13 @@ class debug_panel : public ui_panel {
         m_eventqueuelines,
         m_powergridlines,
+ #ifdef WITH_UART
+        m_rendererlines,
```
Brzydko tak wsadzać pod ifdefa niezwiązane parametry. W ogóle można to zostawić na stałe. Wyłącznik WITH_UART jest żeby dało się kompilować bez libserialport, jak zostanie kilka nieużywanych stringów to nic się nie stanie.

marcinn replied:
ok

Milek7 commented on uart.cpp:
```diff
@@ -158,11 +262,55 @@ uart_input::recall_bindings() {
+
+    if (
+        (!port && std::chrono::duration<float>(now - last_port_find).count() > 1.0)
+        || (port && std::chrono::duration<float>(now - last_port_find).count() > 5.0)
```
Dlaczego enumeruje porty jak jest podłączony?

marcinn replied:
Realizuje use case, w którym podłączasz dodatkowe urządzenie na które chciałbyś się przepiąć. Bez tego trzeba byłoby wyłączać komunikację. Średnio mi się to podoba (stąd rzadszy polling) i wolałbym to mieć na eventach, tyle że wtedy dochodzi zrobienie dedykowanego wątku nasłuchującego. Nie bardzo chciałem wchodzić w threading, ale jak trzeba to rozgryzę temat. 

EDIT: Hm, dałbym rękę uciąć że widziałem informację związaną z eventami i listowaniem portów, ale w libie widzę tylko eventy dla RX, TX i errors. W takim razie mogę zaproponować button "Refresh ports" i przy nawiązanym połączeniu trzeba byłoby go cisnąć, a przy braku połączenia zostawić polling. Albo nawet zostać tylko z tym buttonem. Przełączanie urządzeń odbywa się rzadko, a polling kosztuje.

Milek7 replied:
myślę że z guzikiem będzie ok

Milek7 commented on uart.cpp:
```diff
@@ -7,59 +7,163 @@
+        status->reset_stats();
+        status->is_connected = true;
+    }
+
```
Trzeba też resetować data_pending przy każdym połączeniu.

marcinn replied:
ok

Milek7 commented:
Zrobisz commita z poprawkami? (i rebase na obecny kod)

marcinn commented:
Tak, przysiądę w święta i zrobię.

marcinn committed:
(uart, driveruipanels) always display ui combos for port and baud rate

marcinn committed:
(uart, debug) first part of fixes after code review

- remove `uart_status` from eu7_application class
- remove copy&pasted `m_scenariopanel` size settings
- remove unnecessary code `debug_panel::render_section_uart()`
- remove #ifdef from `driveruipanels.h` textlines
- set `data_pending` to false when establishing a new connection
